torchtune/modules/peft/lora.py

```diff
@@ -130,6 +130,162 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+        activation_qat_config: Optional["FakeQuantizeConfig"] = None,
+        weight_qat_config: Optional["FakeQuantizeConfig"] = None,
+    ):
+        super().__init__(
+            in_dim,
+            out_dim,
+            rank,
+            alpha,
+            dropout=0.0,
+            use_bias=False,
+            quantize_base=False,
+        )
+
+        try:
+            from torchao.quantization.qat.api import FakeQuantizeConfig
+            from torchao.quantization.qat.fake_quantizer import FakeQuantizer
+        except ImportError as err:
+            raise ValueError(
+                "QATLoRALinear is only compatible with torchao 0.7+"
+            ) from err
+
+        # initialize activation fake quantizer
+        if activation_qat_config is not None:
+            assert isinstance(activation_qat_config, FakeQuantizeConfig)
+            self.activation_fake_quantizer = FakeQuantizer(activation_qat_config)
+        else:
+            self.activation_fake_quantizer = None
+
+        # initialize weight fake quantizer
+        if weight_qat_config is not None:
+            assert isinstance(weight_qat_config, FakeQuantizeConfig)
+            group_size = weight_qat_config.group_size
+            if group_size is not None and in_dim % group_size != 0:
+                raise ValueError(
+                    "in_dim (%s) must be divisible by group_size (%s)"
+                    % (in_dim, group_size)
+                )
+            self.weight_fake_quantizer = FakeQuantizer(weight_qat_config)
+        else:
+            self.weight_fake_quantizer = None
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Args:
+            x (torch.Tensor): input tensor with shape ``(..., in_dim)``
+
+        Returns:
+            torch.Tensor: output tensor with shape ``(..., out_dim)``
+
+        """
+        if self.activation_fake_quantizer is not None:
+            _x = self.activation_fake_quantizer(x)
+        else:
+            _x = x
+        if self.weight_fake_quantizer is not None:
+            w = self.weight_fake_quantizer(self.weight)
+        else:
+            w = self.weight
+        out = F.linear(_x, w)
+        if self.disabled:
+            return out
+        lora_out = self.lora_a(self.dropout(x))
+        lora_out = (self.alpha / self.rank) * self.lora_b(lora_out)
+        return out + lora_out
+
+    @classmethod
+    def from_lora_linear(
+        cls,
+        lora_linear: LoRALinear,
+        # TODO: make the types Optional[FakeQuantizeConfig] once we
+        # support torchao 0.7+ by default
+        activation_qat_config: Optional["FakeQuantizeConfig"] = None,
+        weight_qat_config: Optional["FakeQuantizeConfig"] = None,
+    ) -> "QATLoRALinear":
+        """
+        Create a `QATLoRALinear` from an existing `LoRALinear`,
+        preserving the weights and adapters.
+        """
+        if lora_linear.bias is not None:
+            ValueError("Bias is not supported in QAT yet")
```

Comment:
nit: be a bit more explicit in this error message
```suggestion
            ValueError("Bias is not supported in QAT + LoRA yet")
```